### ISSUE TITLE
record values allows arbitrary field names

### ISSUE THREAD

**Issue by bichhawat** [FStarLang/FStar#769]
For record types, when creating a new record using "with" allows arbitrary fields if the first field name following "with" is valid
```
type nstype = {fn:string; ln:string;}
(* works *)
let a = {fn="hello";ln="world"}
let b = {a with fn="go";lname="universe"}
(* fails *)
let a = {fn="hello";ln="world"}
let b = {a with fname="go";ln="universe"}
```

**Comment by bichhawat** [comment]
There is a more general problem:
Record values allow arbitrary field names if the first name is valid when defining values
``
type nstype = {fn:string;}
(* all of the following works *)
let a = {fn="hello";ln="world"}
let b = {fn="go";lname="universe"}
``

**Comment by nikswamy** [comment]
oops. this is an embarrassing bug. Marking it prio 1.

**Comment by tahina-pro** [comment]
206e5b7 actually makes the following code fail:

```
module A
type u = { a : unit ; b : unit }
type v = { a : unit ; c : unit }
let x = { b = () ; a = () }
```

whereas it should succeed, following the fact that the first field of a record term, record pattern or `with` clause should determine the record, so that shadowed fields should still be allowed in the remaining part of the record term, record pattern or `with` clause.

**Comment by tahina-pro** [comment]
6fe50db should solve this.

**Comment by tahina-pro** [comment]
Merged into master along with fix to issue #778 .
